.gitbook/contents/hc/3_write-smart-contract.md

```diff
@@ -0,0 +1,191 @@
+
+```solidity
+  function countFail() public pure {
+      revert("count failed");
+  }
+
+  function justemit() public {
+      emit CalledFrom(msg.sender);
+  }
+
+  event CalledFrom(address sender);
+
+  //helper method to waste gas
+  // repeat - waste gas on writing storage in a loop
+  // junk - dynamic buffer to stress the function size.
+  mapping(uint256 => uint256) public xxx;
+  uint256 public offset;
+
+  function gasWaster(uint256 repeat, string calldata /*junk*/) external {
+      for (uint256 i = 1; i <= repeat; i++) {
+          offset++;
+          xxx[offset] = i;
+      }
+  }
+```
+
+The `HybridAccount` contract has been previously registered to provide access to the `addsub2()` function on our off-chain function. We'll explain more about this later.
+
+## Calling Off-Chain
+
+As already mentioned in the previous section, our off-chain server maps the request made by the bundler via the hashed representation of our function-signature. Let's decode the function-signature we want to call on the off-chain server:
+
+```solidity
+bytes memory req = abi.encodeWithSignature("addsub2(uint32,uint32)", a, b);
+bytes32 userKey = bytes32(abi.encode(msg.sender));
+(uint32 error, bytes memory ret) = HA.CallOffchain(userKey, req);
+
+require(result == HC_ERR_NONE, "Offchain call failed");
+(x,y) = abi.decode(ret,(uint256,uint256)); // x=(a+b), y=(a-b)
+```
+
+{% hint style="info" %}
+The above code block presents us with our first encounter of the acronym "HC". This stands for Hybrid Compute.
+{% endhint %}
+
+We then generate a `userKey` by encoding `msg.sender`. The `userKey` parameter is used to distinguish requests so that they may be processed concurrently without interefering with each other.
+
+Within the Hybrid Account contract itself, the `CallOffchain()` method calls through to another system contract named `HCHelper`:
+
+``` solidity
+function CallOffchain(bytes32 userKey, bytes memory req) public returns (uint32, bytes memory) {
+   require(PermittedCallers[msg.sender], "Permission denied");
+   IHCHelper HC = IHCHelper(_helperAddr);
+   userKey = keccak256(abi.encodePacked(userKey, msg.sender));
+   return HC.TryCallOffchain(userKey, req);
+}
+```
+
+In this example, the `HybridAccount` implements a simple whitelist of contracts which are allowed to call its methods. It would also be possible for a `HybridAccount` to implement additional logic here, such as requiring payment of an `ERC20` token to perform an off-chain call. Conversely, the owner of a `HybridAccount` could choose to make the `CallOffchain()` method available to all callers without restriction.
+
+There is an opportunity for a `HybridAccount` contract to implement a billing system here, requiring a payment of `ERC20` tokens or some other mechanism of collecting payment from the calling contract. This is optional.
+
+## Implement a Helper Contract
```

Comment:
This section is describing what happens within the system-wide HCHelper contract. Developers don't have to implement a Helper contract in this version of Hybrid Compute (although they did in the old system).